pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -109,6 +112,16 @@ public void initialize(ServiceConfiguration conf, ConfigurationCacheService conf
                         log.debug("Policies node couldn't be found for topic : {}", topicName);
                     }
                 } else {
+                    // check if role is authorize to access subscription. (skip validatation if authorization list is empty)     
+                    Set<String> roles = policies.get().auth_policies.subscription_auth_roles.get(subscription);
+                    if (roles != null && !roles.isEmpty() && !roles.contains(role)) {
```

Comment:
Even if the role is superuser, access will be denied unless it is included in subscription_auth_roles, right?

Reply:
> Even if the role is superuser, access will be denied

No, checkout [PersistentTopicsBase:: validateAdminAccessForSubscriber()](https://github.com/apache/pulsar/blob/775894a9f1738b00c8af609494e277203b0a04f6/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java#L241), broker allows super-user/tenant to access any topic api and only validate on subscription if user is not super-user/tenant.

Reply:
Since `ServerCnx` etc calls `AuthorizationService#canConsumeAsync()` directly, superuser seems to fail to subscribe.
```sh
$ grep '^superUserRoles=' conf/standalone.conf

superUserRoles=superuser

$ ./bin/pulsar-admin namespaces policies auth/standalone/test

{
  "auth_policies" : {
    "namespace_auth" : {
      "ns-user" : [ "produce", "consume" ]
    },
    "destination_auth" : { },
    "subscription_auth_roles" : {
      "sub1" : [ "massakam" ]
    }
  },
  ...
}

$ ./bin/pulsar-client consume -s sub1 -n 10 persistent://auth/standalone/test/t1
```
```
17:35:48.140 [pulsar-io-49-3] INFO  org.apache.pulsar.broker.service.ServerCnx - New connection from /127.0.0.1:43666
17:35:48.162 [pulsar-io-49-3] INFO  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:43666] Client successfully authenticated with basic role superuser and originalPrincipal null
17:35:48.182 [pulsar-io-49-3] WARN  org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider - [superuser] is not authorized to subscribe on persistent://auth/standalone/test/t1-sub1
17:35:48.184 [pulsar-io-49-3] WARN  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:43666] Client is not authorized to subscribe with role superuser
```
https://github.com/apache/pulsar/blob/0da4e4af407650de874ed6c9de891d98aa540aad/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L542-L548

Reply:
@rdhabalia Now the super user can subscribe, but consumers which are allowed access at the level of namespace, topic, or subscription fails to subscribe yet.

Reply:
@massakam yes, that's correct. that's what we want here right?
- first broker validates subscription-permission (if it's configured) over namespace-auth so, only authorized principal can access the subscription. and that will also cover "pre-fix" sub usecase where only authorized sub can consume on specific subscription.

@merlimat can you also review this PR one more time.